### PR TITLE
pow2: more implementations

### PR DESCRIPTION
**Pull request by yoni206**
This PR adds implementations of functions from the `pow2` solver, rewriter and type checker.

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -381,6 +383,21 @@ RewriteResponse ArithRewriter::postRewriteMult(TNode t){
...
 }
 
+RewriteResponse ArithRewriter::postRewritePow2(TNode t)
+{
+  Assert(t.getKind() == kind::POW2);
+  NodeManager* nm = NodeManager::currentNM();
+  // if constant, we eliminate
+  if (t[0].isConst())
+  {
+    Integer i = t[0].getConst<Rational>().getNumerator();
```
Comment (nafur) [inline src/theory/arith/arith_rewriter.cpp:393]: What if it is rational?

File: src/theory/arith/nl/pow2_solver.cpp [hunk 2]
```diff
@@ -61,11 +61,81 @@ void Pow2Solver::initLastCall(const std::vector<Node>& assertions,
...
+      continue;
+    }
+    d_initRefine.insert(i);
+    // initial refinement lemmas
+    std::vector<Node> conj;
+    // x>=0 -> x < pow2(x)
+    Node xgeq0 = nm->mkNode(LEQ, d_zero, i[0]);
+    Node xltpow2x = nm->mkNode(LT, i[0], i);
+    conj.push_back(nm->mkNode(IMPLIES, xgeq0, xltpow2x));
+    Node lem = conj.size() == 1 ? conj[0] : nm->mkNode(AND, conj);
```
Comment (nafur) [inline src/theory/arith/nl/pow2_solver.cpp:82]:
```suggestion
    Node lem = nm->mkAnd(AND, conj);
```

File: src/theory/arith/arith_rewriter.cpp (hunk 1 above)
Comment (yoni206) [reply src/theory/arith/arith_rewriter.cpp:395]: Then it should fail on type checking, [here](https://github.com/cvc5/cvc5/pull/6756/files#diff-c38305abf5d7b3f2e1ad0c47b5525b14721ba385967d30012a9cfcac0cc1ae5bR146) no? Anyway, I've added an assertion in 84b524d.

File: src/theory/arith/nl/pow2_solver.cpp [hunk 3]
```diff
@@ -61,11 +61,81 @@ void Pow2Solver::initLastCall(const std::vector<Node>& assertions,
...
+    conj.push_back(nm->mkNode(IMPLIES, xgeq0, xltpow2x));
+    Node lem = nm->mkAnd(conj);
+    Trace("pow2-lemma") << "Pow2Solver::Lemma: " << lem << " ; INIT_REFINE"
+                        << std::endl;
+    d_im.addPendingLemma(lem, InferenceId::ARITH_NL_POW2_INIT_REFINE);
+  }
+}
 
-void Pow2Solver::checkFullRefine() {}
+void Pow2Solver::checkFullRefine()
```
Comment (ajreynol) [inline src/theory/arith/nl/pow2_solver.cpp:89]: FYI, one lemma to consider eventually is x <= y => pow2(x) <= pow2(y)

**Review by ajreynol — APPROVED** [review]

File: src/theory/arith/nl/pow2_solver.cpp (hunk 3 above)
Comment (yoni206) [reply src/theory/arith/nl/pow2_solver.cpp:89]: Great, I'll add this to the next PR

**Pull request merged**